src/ray/gcs/gcs_server/gcs_node_manager.h

```diff
@@ -129,7 +129,7 @@ class GcsNodeManager : public rpc::NodeInfoHandler {
 
   /// Drain the given node.
   /// Idempotent.
-  void DrainNode(const NodeID &node_id);
+  virtual void DrainNode(const NodeID &node_id);
```

Comment:
Why do we make it virtual? Is it only for testing? We shouldn't do this only for testing purpose.

Reply:
hmm i thought a bit about this, but given just the language constraints of c++ there seem to be 2 main ways:
1. Do `MockGcsNodeManager : rpc::NodeInfoHandler` or some other virtual class (instead of extending `GcsNodeManager`). I'm not sure what that buys us though (since there would still be 2 layers of indirection).
2. The way gMock [recommends](https://github.com/google/googletest/blob/main/docs/gmock_cook_book.md#mocking-non-virtual-methods-mockingnonvirtualmethods) is to use templating for high performance code, but I don't think we fall in this category.